design/cassandra/sidecar.md

```diff
@@ -0,0 +1,79 @@
+## Sidecar Design Proposal
+
+### Consideration: REST API
+
+When thinking about how our sidecar will communicate with our controller, a natural solution that comes to mind is though a REST API. The sidecar will run an HTTP Server which the other party will call. This is the approach used by [Netflix's Priam](https://github.com/Netflix/priam/wiki).
+
+However, this includes some extra complications. Remember that, according to our goals we are designing a level-based system.
+First of all, some operations just take a long time. Backup, for example, might take hours to complete. 
+That means our operator must have an open TCP connection for all this time. If it gets interrupted, which we do expect to happen, this connection will be lost and we won't have any record that it ever happened. 
+
+
+This doesn't seem like the Kubernetes way of doing things. Consider this example, as one could think of the kubelet as a sidecar for Pods: 
+
+* **Question:** Does the scheduler ping the kubelet each time it schedules a Pod and then wait for an answer? 
+* **Answer:** No, it writes the `nodeName` field on the PodSpec. In other words, it writes a record of intent. No matter how many times the kubelet or the scheduler crashes it doesn't matter. The record of intent is there. 
+
+### Control-Loop Design
+
+* Based on our observations above, we design a method of communication in line with the Kubernetes philosophy.
+* When the controller wants to invoke a functionality in a Sidecar, it should write a record of intent in the Kubernetes API Objects (etcd). The sidecar will be watching the Kubernetes API and responding accordingly.
+* There are two approaches to represent the record of intent:
+  1. **Labels:**
+    * When the controller wants to communicate with a sidecar, it will write a predefined label in the ClusterIP Service Object of the specific instance. For example, to communicate that we want an instance to terminate, we could write the label 'cassandra.rook.io/termination-timestamp=0`. The sidecar will see this and decommission the Cassandra instance. When it is done, it will change the label value to a valid timestamp. Then the controller will know to delete that instance.
```

Comment:
That's not really defined somewhere atm.
I will add documentation as comments in the `pkg/operator/cassandra/constants` package.
It may also be useful to document them in a separate file.